zfp: several package improvements

lindstro opened the pull request:
- add variants for build targets, language bindings, backends
- ensure selected variants are compatible with zfp version
- point to GitHub (not LLNL) tar balls
- add dependencies
- update link to homepage
- add maintainers

lindstro committed:
zfp: several package improvements

- add variants for build targets, language bindings, backends
- ensure selected variants are compatible with zfp version
- point to GitHub (not LLNL) tar balls
- add dependencies
- update link to homepage
- add maintainers

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -6,26 +6,45 @@
+    homepage    = 'https://zfp.llnl.gov'
+    url         = 'https://github.com/LLNL/zfp/releases/download/0.5.5/zfp-0.5.5.tar.gz'
+    git         = 'https://github.com/LLNL/zfp.git'
+    maintainers = ['lindstro', 'GarrettDMorrison']
```
:+1:

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -6,26 +6,45 @@
+    version('0.5.3', sha256='a5d2f8e5b47a7c92e2a5775b82cbfb3a76c87d0ac83d25abb4ac10ea75a2856e')
+    version('0.5.2', sha256='9c738ec525cc76b4bb80b2b3f7c9f07507eeda3a341470e5942cda97efbe9a4f', url='https://github.com/LLNL/zfp/archive/0.5.2/zfp-0.5.2.tar.gz')
+    version('0.5.1', sha256='f255dd1708c9ae4dc6a56dd2614e8b47a10d833c87fd349cbd47545a19c2b779', url='https://github.com/LLNL/zfp/archive/0.5.1/zfp-0.5.1.tar.gz')
+    version('develop', branch='develop')
```
Can we keep this version at the top? It's common practice throughout the packages to list versions from highest to lowest and for Spack's comparison semantics `develop` count as `+inf`.

lindstro replied:
Of course.

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -6,26 +6,45 @@
+    variant('shared',    default=True,  description='Build shared libraries')
+
+    # Language bindings
+    variant('c',       default=False, description='Enable C bindings')
+    variant('python',  default=False, description='Enable Python bindings')
+    variant('fortran', default=False, description='Enable Fortran bindings')
```
Just a comment, not a request. If it makes sense for this package these three Boolean variants can be also written as a single multi-valued variant:
```suggestion
    variant(
        'bindings', values=any_combination_of('c', 'python', 'fortran'),
        description='Language bindings to enable'
    )
```

lindstro replied:
I think that would be fine.  In the future, we may add bindings for different components of zfp, in which case we'd have to be more specific about what each variant enables.  But I guess that could be handled similarly.  I'm fine with making this change.

lindstro replied:
In the end, I decided against this.  We have similar Boolean variants for build targets, backends, internal configuration parameters, etc.  Some combinations of these are not as straightforward to categorize.

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -38,15 +57,26 @@ class Zfp(CMakePackage):
 
+        # Common CMake options for zfp 0.5.1 and later
         args = [
+            '-DBUILD_UTILITIES:BOOL={0}'.format(
+                'ON' if '+utilities' in spec else 'OFF'),
```
All of these CMake defines can be shortened using convenience methods. For instance the one above can be written as:
```suggestion
           self.define_from_variant('BUILD_UTILITIES', 'utilities'),
```

lindstro replied:
I like it!

lindstro replied:
One note: I just realized that our CMake scripts currently don't install the utilities, and for spack it doesn't make sense to build the examples.  We'll fix this in the next zfp release, so maybe we can leave this line in but commented out?

lindstro replied:
Quick question on how to use this with testing (please forgive my limited knowledge of Spack, CMakePackage, and Python).  Currently, the package lists this: `'-DBUILD_TESTING:BOOL={0}'.format('ON' if self.run_tests else 'OFF')`.  However, I also added a variant `tests` with the intent that this needs to be enabled to build the tests.  In other words, I would want to rewrite this as `self.define_from_variant('BUILD_TESTING', 'tests')`, but I understand that `self.run_tests` is set on the command-line via the `--run-tests` Spack option.  Would you advise simply doing away with the tests variant and relying on the Spack option to enable testing?  I'm guessing something like `self.define('BUILD_TESTING', self.run_tests)` would do the trick?

Similarly, I would want to replace `'-DZFP_BIT_STREAM_WORD_SIZE:STRING={0}'.format(spec.variants['bsws'].value)` with `self.define('ZFP_BIT_STREAM_WORD_SIZE', spec.variants['bows'].value)`.  Any concern here?

alalazo replied:
> I'm guessing something like self.define('BUILD_TESTING', self.run_tests) would do the trick

That's correct. The preference would be to turn all the dependencies needed for testing into "test" type dependencies, and rely on `self.run_test`. 

> Similarly, I would want to replace '-DZFP_BIT_STREAM_WORD_SIZE:STRING={0}'.format(spec.variants['bsws'].value) [ ... ] Any concern here?

I'm absolutely fine with that change!

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -55,15 +85,64 @@ def cmake_args(self):
+
+        # cfp compressed-array C bindings
+        if '+c' in spec:
+            if not spec.satisfies('@0.5.4:'):
+                raise InstallError('+c requires zfp 0.5.4 or later')
```
Rather than raising `InstallError`, which would fail at build time after having installed all the build dependencies, you can use a `conflicts` directive:
```
conflicts('+c', when='@:0.5.3', msg='+c requires zfp 0.5.4 or later')
```
The information from a `conflicts` directive is used at concretization time either to validate a solution and eventually fail with the given error message (original concretizer), or to reduce the search space for a concretized spec (ASP based concretizer).

lindstro replied:
Nice!  Yes, let's go with that.

lindstro replied:
Although if we go with a 'bindings' variant, this would have to be reworked a bit.

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -6,26 +6,45 @@
-       lossy compression of floating-point and integer multi-dimensional
-       arrays.
+class Zfp(CMakePackage, CudaPackage):
+    """zfp is a compressed format for representing multidimensional
+       floating-point and integer arrays. zfp provides compressed-array
+       classes that support high throughput read and write random access
+       to individual array elements. zfp also supports serial and parallel
+       (OpenMP and CUDA) compression of whole arrays.
```
The first paragraph in a package docstring is used by Spack wherever a short description is needed, while the entire docstring is used as a long description. I would thus suggest to start a new paragraph after the first sentence:
```suggestion
    """zfp is a compressed format for representing multidimensional
       floating-point and integer arrays. 
       
       zfp provides compressed-array
       classes that support high throughput read and write random access
       to individual array elements. zfp also supports serial and parallel
       (OpenMP and CUDA) compression of whole arrays.
```

lindstro replied:
Works for me.

alalazo reviewed (changes requested):
Thanks for the PR and signing up as a maintainer. I left a few comments that hopefully would help improve the package recipe. Let me know what you think.

lindstro committed:
zfp: address suggestions by Spack team

- use conflicts() instead of raising exceptions
- use define() and define_from_variant() where applicable

alalazo commented:
I think something went wrong during the last push...

lindstro commented:
> I think something went wrong during the last push...

Woah--not sure what happened here.  I fetched from upstream and then rebased my latest commit on top.  Not sure why all those other commits are showing up here, but I'm still far from a git expert and have no experience with pushing while in the middle of a PR.

Before I make an even bigger mess, I figured I'd get your opinion.  How about I `git reset --hard` back to my commit (9598b94) before the rebase  and then force push this single commit to my fork?  There should be no conflicts here.

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -6,28 +6,48 @@
+    version('0.5.1', sha256='f255dd1708c9ae4dc6a56dd2614e8b47a10d833c87fd349cbd47545a19c2b779', url='https://github.com/LLNL/zfp/archive/0.5.1/zfp-0.5.1.tar.gz')
+
+    # Build targets
+    # variant('utilities', default=True,  description='Build utilities')
```
```suggestion
    # TODO: variant('utilities', default=True,  description='Build utilities')
```

alalazo commented on var/spack/repos/builtin/packages/zfp/package.py:
```diff
@@ -38,32 +58,47 @@ class Zfp(CMakePackage):
-                'ON' if '+shared' in spec else 'OFF'),
-            '-DBUILD_TESTING:BOOL={0}'.format(
-                'ON' if self.run_tests else 'OFF')
+            # self.define_from_variant('BUILD_UTILITIES', 'utilities'),
```
```suggestion
            # TODO: self.define_from_variant('BUILD_UTILITIES', 'utilities'),
```

alalazo reviewed (commented):
Thanks for the changes, this looks great! I have just a minor request to add `TODO` in front of commented code (to make it clear the code is there on purpose). Otherwise I think this is ready to be merged.

lindstro committed:
Apply suggestions from code review

Co-authored-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>
